### PR TITLE
adds inline ability for es6 output

### PR DESCRIPTION
**Pull request by pflannery**
Adds a new transformer module called `InlineES6ModuleTransformer` for in-lining ES6 imports (as parts) with a root module.

usage: ```traceur --modules=inline --outputLanguage=es6 --out ./out/some-lib.js ./src/root.js```

transforms:

```js
 // a.js
 export function a() {
   return 'A';
 }

 // root.js
 import {a} from './a.js';
 export function b() {
   return a();
 }
```

into:

```js
 const $__module_a_js__ = (function() {
   "use strict";
   function a() {
     return 'A';
   }
   return {get a() {
       return a;
     }};
 })();
 "use strict";
 const {a} = $__module_a_js__;
 export function b() {
   return a();
 }
```

### CODE REVIEW

**Comment by pflannery** [comment]
relates with #1842
blocked from running tests on windows because of https://github.com/google/traceur-compiler/pull/1832

**Comment by arv** [comment]
Thanks!

Can this be tested? Maybe using a unit test.

**Comment by pflannery** [comment]
@arv yes once I've figured out how to get the tests to run on windows. dont merge yet.

Is there a way to determine if the module being transformed is the top dominator? because when specifying a single source file in the cli it would be cool if it we could preserve expose it's es6 export keywords.

**Comment by arv** [comment]
Is it `make` that is causing you trouble?

You can also run the tests directly so that you can skip the currently broken tests on windows.

**Comment by pflannery** [comment]
its failing [here](https://github.com/google/traceur-compiler/pull/1832#issuecomment-87713400) and without the tval working it cant run any es6 test..

**Comment by pflannery** [comment]
@arv I've added a test but not sure how to actually get the import (`test_a.js`, `test_b.js`) contents injected in to the result, any advice? 

and also before we merge this would be nice to be able to determine the root dominator file so we can preserve the ES6 export keywords for import to another ES6 library..

**Comment by arv** [comment]
@pflannery Just to be clear. Is your use case that you want to compile a Module and its dependencies into a single ES6 Module. Something like:

Input:
```js
// a.js
export function a() {
  return 'A';
}
```
```js
// root.js
import {a} from './a.js';
export function b() {
  return a();
}
```
Outputs:
```js
// root-compiled.js
var $tmp = (function() {
  "use strict";
  function a() {
    return 'a';
  }
  return {a};
})();
const {a} = $tmp;
export function b() {
  return a();
}
```

**Comment by pflannery** [comment]
@arv yes

**Comment by pflannery** [comment]
@arv I've separated this transform in to a separate transformer module called `InlineES6ModuleTransformer` and updated the tests..

transforms:

```js
 // a.js
 export function a() {
   return 'A';
 }

 // root.js
 import {a} from './a.js';
 export function b() {
   return a();
 }
```

into:

```js
 const $__module_a_js__ = (function() {
   "use strict";
   function a() {
     return 'A';
   }
   return {get a() {
       return a;
     }};
 })();
 "use strict";
 var a = ($__module_a_js__).a;
 export function b() {
   return a();
 }
```

**Comment by pflannery** [comment]
@arv PTAL

If good to go then I will rebase to current master and squash ;) thanks

File: src/codegeneration/InlineES6ModuleTransformer.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+  createVariableStatement,
+  createUseStrictDirective
+  } from './ParseTreeFactory.js';
+import globalThis from './globalThis.js';
+import scopeContainsThis from './scopeContainsThis.js';
+import {IMPORT_SPECIFIER_SET} from '../syntax/trees/ParseTreeType.js';
+
+let anonInlineModules = 0;
+
+// transforms:
```
Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: Make this into a single comment.

Can you also provide some information what the usage of this is?

File: src/codegeneration/InlineES6ModuleTransformer.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+  // gives the module a unique name
+  transformModuleSpecifier(tree) {
+    return createBindingIdentifier(this.getTempVarNameForModuleSpecifier(tree));
+  }
+
+  // tests if we are a root module or not
+  isRootModule() {
+    let rootModules = this.options_.rootModules || [];
+    return rootModules.some((testModule)=> {
```
Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: ws betwen `)` and `=>`

File: src/codegeneration/InlineES6ModuleTransformer.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    let initializer = this.transformAny(tree.moduleSpecifier);
+    return createVariableStatement(CONST, binding, initializer);
+  }
+
+  // gives the module a unique name
+  transformModuleSpecifier(tree) {
+    return createBindingIdentifier(this.getTempVarNameForModuleSpecifier(tree));
+  }
+
+  // tests if we are a root module or not
```
Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: You can skip this comment. The method name provides the same information

File: src/codegeneration/InlineES6ModuleTransformer.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    return rootModules.some((testModule)=> {
+      return this.moduleName === testModule.name;
+    });
+  }
+
+  getTempVarNameForModuleName(moduleName) {
+    return '$__module__' +
+      moduleName
+      .split(/[^a-zA-Z0-9$]/g)
+      .filter((value) => {return value;})
```
Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: .filter((value) => value})

File: src/codegeneration/PureES6Transformer.js [hunk 5]
```diff
@@ -63,5 +64,11 @@ export class PureES6Transformer extends MultiTransformer {
     }
     append(AnnotationsTransformer);
     append(TypeTransformer);
+
+    // inline if the option was specified
```
Comment (arv) [inline src/codegeneration/PureES6Transformer.js]: remove comment

File: src/node/recursiveModuleCompile.js [hunk 6]
```diff
@@ -50,6 +50,8 @@ function recursiveModuleCompileToSingleFile(outputFile, includes, options) {
     return include;
   });
 
+  options.rootModules = resolvedIncludes;
```
Comment (arv) [inline src/node/recursiveModuleCompile.js]: I wonder if we could pass this information in some other way? It feels a bit odd to put this on the options object.

File: test/unit/codegeneration/PureES6Transformer.js [hunk 7]
```diff
@@ -48,4 +48,44 @@ suite('PureES6Transformer.js', function() {
     var transformed = transformer.transform(tree);
     assert.equal(write(transformed), write(expectedTree));
   });
+
+  test('Inline', function() {
+    var options = new Options({
```
Comment (arv) [inline test/unit/codegeneration/PureES6Transformer.js]: Why the getters?

```js
var options = new Options({
  modules: 'inline',
  rootModules: {name: null}
})
```

File: test/unit/codegeneration/PureES6Transformer.js [hunk 8]
```diff
@@ -48,4 +48,44 @@ suite('PureES6Transformer.js', function() {
...
+      get modules() { return "inline"; },
+      get rootModules() { return [{name: null}]; }
+    });
+    var reporter = new ErrorReporter();
+
+    var code = [
+      "import {TestA} from './resources/test_a.js';",
```
Comment (arv) [inline test/unit/codegeneration/PureES6Transformer.js]: We use single quotes everywhere... but in this case you can use a template literal.

File: src/codegeneration/InlineES6ModuleTransformer.js [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+
+  // tests if we are a root module or not
+  isRootModule() {
+    let rootModules = this.options_.rootModules || [];
+    return rootModules.some((testModule)=> {
+      return this.moduleName === testModule.name;
+    });
+  }
+
+  getTempVarNameForModuleName(moduleName) {
```
Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: It is not clear why we cannot use the existing logic here?

File: src/codegeneration/InlineES6ModuleTransformer.js [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+ *  const {a} = $__module__a__js;
+ *  export function b() {
+ *    return a();
+ *  }
+ */
+export class InlineES6ModuleTransformer extends ModuleTransformer {
+
+  // override the module entry point to place "use strict" at the top
+  moduleProlog() {
+    return [createUseStrictDirective()];
```
Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: Modules are always strict so adding this is just noise.

File: src/codegeneration/InlineES6ModuleTransformer.js [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+  }
+
+  // override to wrap module in a closure unless it's a root module
+  wrapModule(statements) {
+    let seed = this.moduleName || 'anon_' + ++anonInlineModules;
```
Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: It is not clear to me when we do not have a module name here?

**Comment by arv** [comment]
Very nice.

Can you look at the Travis error too?

File: src/codegeneration/InlineES6ModuleTransformer.js (hunk 11 above)
Comment (pflannery) [inline src/codegeneration/InlineES6ModuleTransformer.js]: me too, I've just reused the existing inline transformer logic [from here](https://github.com/google/traceur-compiler/blob/master/src/codegeneration/InlineModuleTransformer.js#L38)

File: src/codegeneration/InlineES6ModuleTransformer.js (hunk 10 above)
Comment (pflannery) [inline src/codegeneration/InlineES6ModuleTransformer.js]: I should be a bit clearer here, it prevents the default ```var __moduleName = "../some/path/to/module.js";``` being generated and for every module

File: src/codegeneration/PureES6Transformer.js (hunk 5 above)
Comment (pflannery) [inline src/codegeneration/PureES6Transformer.js]: will do this and reformat other comments to use doc comments `/**`

File: src/codegeneration/InlineES6ModuleTransformer.js (hunk 9 above)
Comment (pflannery) [inline src/codegeneration/InlineES6ModuleTransformer.js]: will remove

File: src/codegeneration/InlineES6ModuleTransformer.js (hunk 10 above)
Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: Maybe just return `[]`?

File: src/codegeneration/InlineES6ModuleTransformer.js (hunk 11 above)
Comment (johnjbarton) [inline src/codegeneration/InlineES6ModuleTransformer.js]: I guess only roots of the dependency tree can be anonymous, eg the ones coming via `Loader.module()`.

File: src/codegeneration/InlineES6ModuleTransformer.js (hunk 10 above)
Comment (pflannery) [inline src/codegeneration/InlineES6ModuleTransformer.js]: just trying this and it removes 'use strict' all together

File: src/node/recursiveModuleCompile.js (hunk 6 above)
Comment (pflannery) [inline src/node/recursiveModuleCompile.js]: Maybe via metadata but it will need to be passed to the transformer too. There is a fair amount of complex workflow to get there..

File: src/codegeneration/InlineES6ModuleTransformer.js (hunk 10 above)
Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: Isn't that what we want?

File: src/node/recursiveModuleCompile.js (hunk 6 above)
Comment (arv) [inline src/node/recursiveModuleCompile.js]: @johnjbarton Do you have any suggestion here?

File: src/codegeneration/InlineES6ModuleTransformer.js (hunk 10 above)
Comment (pflannery) [inline src/codegeneration/InlineES6ModuleTransformer.js]: if that is the case then would it need to put one "use strict" at the top of the output or does "use strict" hoist?

Comment (arv) [inline src/codegeneration/InlineES6ModuleTransformer.js]: The output of this transformer is a Module which is already strict.

File: src/node/recursiveModuleCompile.js (hunk 6 above)
Comment (johnjbarton) [inline src/node/recursiveModuleCompile.js]: My initial reaction was to agree with @pflannery: the better solution is metadata and it is probably better to do that in a separate PR.  Then #1859

Comment (pflannery) [inline src/node/recursiveModuleCompile.js]: :+1:  thanks @johnjbarton

**Comment by pflannery** [comment]
@arv @johnjbarton I've implemented root module detection using metadata and removed code that references the options object. 

There is a case where the metadata is not defined when using the --dir switch, however I'm not sure anyone would ever use '--dir', `inline` and `es6`  together but it does work. Because of this I've made sure that the metadata object is defined when evaluating `isRootModule`.

PTAL

**Comment by johnjbarton** [comment]
Thanks, the metadata approach looks much better.

I'm ok with this but for your sake I think another PR with a test with the command line (generated-code-dependencies.js) would be a good addition. The test would take two modules, generate output and run the resulting file; the root module would just use a value from its dependency. This exercises all of the elements without being effected by the details of the code gen.  Especially in this case the rootModule which is null in the current test.

**Comment by pflannery** [comment]
@johnjbarton no problem. I can add the tests to this PR.

**Comment by pflannery** [comment]
Glad I did this as I found an issue where I needed to replicate what `InlineModuleTransformer` does with `transformNamedExport`
```js
transformNamedExport(tree) {
    return createEmptyStatement();
}
```
There seems to be a little bug with this as an empty line with `;` get generated,  I see this happening with the current InlineModuleTransformer too. 
It doesn't break the tests and is valid runnable code. I believe this also relates to #1822

**Comment by arv** [comment]
@pflannery You can return an AnonBlock instead. Those gets removed/expanded in transformList.

File: src/codegeneration/PureES6Transformer.js [hunk 12]
```diff
@@ -65,5 +66,9 @@ export class PureES6Transformer extends MultiTransformer {
     }
     append(AnnotationsTransformer);
     append(TypeTransformer);
+
+    if (options.modules.toLowerCase() === 'inline') {
```
Comment (arv) [inline src/codegeneration/PureES6Transformer.js]: I would skip the toLowerCase here. No need for us to support case insensitive values here.

**Comment by arv** [comment]
LGTM

Thanks for taking care of this.

File: src/codegeneration/PureES6Transformer.js (hunk 12 above)
Comment (pflannery) [inline src/codegeneration/PureES6Transformer.js]: ok I will update this now

**Comment by pflannery** [comment]
> You can return an AnonBlock instead. Those gets removed/expanded in transformList.

@arv I've switched to use 
```js
transformNamedExport(tree) {
    return new AnonBlock(null, []);
}
```

**Comment by arv** [comment]
Actually, I still don't understand why you need to add the `'use strict'` to the top of your module?

**Comment by pflannery** [comment]
@arv currently with iojs and chrome canary with `--harmony-classes` enabled I'm seeing this error
```js
SyntaxError: Block-scoped declarations (let, const, function, class) not yet supported outside strict mode
```

**Comment by pflannery** [comment]
if there isn't a way around this then I could put a `TODO` to remove this once this condition is lifted?

**Comment by arv** [comment]
Lets leave the use strict for now. It does not hurt.

However, it is not clear what the intended output is? Right now ModuleTransformer transforms a Module into a Script and a Script cannot contain an ExportDeclaration (validation is always off for PureES6Transformer). But what we really want to do is to transform a Module into a Module and execute it as a module, which of course is not possible with V8 today.

**Comment by arv** [comment]
Merging
